### PR TITLE
Hometask7

### PR DESCRIPTION
**Pull request by MihailNenahov**
I still have some questions, so please look at commit history and write some comments about what was wrong.

### CODE REVIEW

File: first_task/hello_world/hello_world.go [hunk 1]
```diff
@@ -1,9 +0,0 @@
-//i'll left some questions
```
Comment (imatsiash-ib) [inline first_task/hello_world/hello_world.go]: Why I can see redundant files in this PR ?

File: task1/main.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	"encoding/json"
+	"log"
+	"net/http"
+)
+
+type answer struct {
+	Host       string      `json:"host"`
+	UserAgent  string      `json:"user_agent"`
+	RequestUri string      `json:"request_uri"`
+	H          http.Header `json:"headers"`
```
Comment (imatsiash-ib) [inline task1/main.go]: H ? What does it mean ?
Please use more obvious names.

File: task1/main.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+func HandleAnsw(w http.ResponseWriter, r *http.Request) {
+	ans := answer{
+		r.Host,
+		r.UserAgent(),
+		r.RequestURI,
+		r.Header,
+	}
+
+	w.Header().Set("Content-Type", "applicatioin/json; charset=UTF-8")
+	buf, _ := json.Marshal(ans)
```
Comment (imatsiash-ib) [inline task1/main.go]: Why do you ignore an error returned by `Marshal` function ?

File: task1/main.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	ans := answer{
+		r.Host,
+		r.UserAgent(),
+		r.RequestURI,
+		r.Header,
+	}
+
+	buf, err := json.Marshal(ans)
+	if err != nil {
+		log.Println(err)
```
Comment (imatsiash-ib) [inline task1/main.go]: In case of error you should return error message to the client, and set status code to an error.

File: task1/main.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	ans := answer{
+		r.Host,
+		r.UserAgent(),
+		r.RequestURI,
+		r.Header,
+	}
+
+	buf, err := json.Marshal(ans)
+	if err != nil {
+		fmt.Fprint(w, err)
```
Comment (wshaman) [inline task1/main.go:28]: Error header should be set here.

File: task2/server/server.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	defer ln.Close()
+	if err != nil {
+		log.Fatal("1) ", err)
+	}
+
+	for {
+		conn, err := ln.Accept()
+		fmt.Println(conn.RemoteAddr().String(), " connected")
+		if err != nil {
+			fmt.Print("2) ")
```
Comment (wshaman) [inline task2/server/server.go]: `log.Panic` does the same.

File: task3/error.html [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html>
+<body>
+
+<h2>404</h2>
+
+<p>my teacher told me to not show sense of error to user, so u just need to know, that smth went wrong</p>
```
Comment (wshaman) [inline task3/error.html:7]: custom 404 is a cool tradition.

File: task3/index.html [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<!DOCTYPE html>
+<html>
```
Comment (wshaman) [inline task3/index.html:2]: any place to show cookie?

File: task3/main.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	http.SetCookie(w, &c)
+}
+
+func handleConn(w http.ResponseWriter, r *http.Request) {
+	if !correctMethod(r) {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		t, _ := template.ParseFiles("error.html")
+		t.Execute(w, errorToSend{
+			http.StatusText(http.StatusMethodNotAllowed),
```
Comment (wshaman) [inline task3/main.go:36]: I got the idea, anyway here should be different mesages

**Review by wshaman — APPROVED** [review]
That's OK. Only one thing isn't working for me, it's an HTML layout in 3rd task, but it's not critical, we are not learning HTML here.

File: task1/main.go (hunk 5 above)
Comment (imatsiash-ib) [reply task1/main.go:28]: Yes, you should return error to the user.

File: task2/client/client.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+	defer conn.Close()
+
+	for {
+		input := bufio.NewReader(os.Stdin)
+
+		fmt.Println("enter data to send")
+
+		data, err := input.ReadString('\n')
+
+		if err != nil {
```
Comment (imatsiash-ib) [inline task2/client/client.go:23]: Please remove all extra empty lines.

Comment (imatsiash-ib) [reply task2/client/client.go:23]: Please do not ignore an error.

File: task2/client/client.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+			break
+		}
+
+		check := strings.TrimSuffix(data, string('\n'))
+
+		if check == "EXIT" {
```
Comment (imatsiash-ib) [inline task2/client/client.go]: What happens if I write `exit` ?

File: task2/server/server.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+		}
+
+	}
+}
+
+func main() {
+	ln, err := net.Listen("tcp", ":8083")
+	defer ln.Close()
+	if err != nil {
+		log.Fatal("1) ", err)
```
Comment (imatsiash-ib) [inline task2/server/server.go]: What `1) ` does mean ?

File: task2/server/server.go [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+func handleAns(conn net.Conn) {
+	defer fmt.Println(conn.RemoteAddr().String(), " closed")
+	defer conn.Close()
+	for {
+		m, err := bufio.NewReader(conn).ReadString('\n')
+		if err != nil {
+			if err != io.EOF {
+				fmt.Println(err)
+			}
+			break
```
Comment (imatsiash-ib) [inline task2/server/server.go:19]: Strange construction..

**Review by burov — APPROVED** [review]

File: task2/client/client.go (hunk 11 above)
Comment (MihailNenahov) [reply task2/client/client.go]: `exit` is not a command for exit. if `exit` is input, so this string will be sent to server and server will process this string and will return `EXIT`.

File: task2/server/server.go (hunk 12 above)
Comment (MihailNenahov) [reply task2/server/server.go]: it helps to find a place, where program felt.

Comment (imatsiash-ib) [reply task2/server/server.go]: This is a debug information, that will be redundant for the client.

File: task2/client/client.go (hunk 11 above)
Comment (imatsiash-ib) [reply task2/client/client.go]: From homework description:
C exits by input `exit`.

I expect that when client input == `exit`, client should be stopped. Am I correct?

Comment (MihailNenahov) [reply task2/client/client.go]: u are right. my fault.

**Review by imatsiash-ib — APPROVED** [review]

**Pull request closed without merging**